x-pack/agent/pkg/fleetapi/actions.go

```diff
@@ -0,0 +1,164 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package fleetapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// Action base interface for all the implemented action from the fleet API.
+type Action interface {
+	fmt.Stringer
+	Type() string
+	ID() string
+}
+
+// BaseAction is the base of all actions to be executed.
+type BaseAction struct {
+	ActionID   string
+	ActionType string
+}
+
+// Type returns the action type.
+func (a *BaseAction) Type() string {
+	return a.ActionType
+}
+
+// ID returns the action ID.
+func (a *BaseAction) ID() string {
+	return a.ActionID
+}
+
+// UnknownAction is an action that is not know by the current version of the Agent and we don't want
+// to return an error at parsing time but at execution time we can report or ignore.
+//
+// NOTE: We only keep the original type and the action id, the payload of the event is dropped, we
+// do this to make sure we do not leak any unwanted information.
+type UnknownAction struct {
+	*BaseAction
+	originalType string
+}
+
+// Type returns the type of the Action.
+func (a *UnknownAction) Type() string {
+	return "UNKNOWN"
+}
+
+func (a *UnknownAction) String() string {
+	var s strings.Builder
+	s.WriteString("action_id: ")
+	s.WriteString(a.ID())
+	s.WriteString(", type: ")
+	s.WriteString(a.Type())
+	s.WriteString(" (original type: ")
+	s.WriteString(a.OriginalType())
+	s.WriteString(")")
+	return s.String()
+}
+
+// OriginalType returns the original type of the action as returned by the API.
+func (a *UnknownAction) OriginalType() string {
+	return a.originalType
+}
+
+// PolicyChangeAction is a request to apply a new
+type PolicyChangeAction struct {
+	*BaseAction
+	Policy map[string]interface{} `json:"policy"`
+}
+
+func (a *PolicyChangeAction) String() string {
+	var s strings.Builder
+	s.WriteString("action_id: ")
+	s.WriteString(a.ID())
+	s.WriteString(", type: ")
+	s.WriteString(a.Type())
+	return s.String()
+}
+
+// Actions is a list of Actions to executes and allow to unmarshal heterogenous action type.
+type Actions []Action
+
+// UnmarshalJSON takes every raw representation of an action and try to decode them.
+func (a *Actions) UnmarshalJSON(data []byte) error {
+	type r struct {
+		ActionType string          `json:"type"`
+		ActionID   string          `json:"id"`
+		Data       json.RawMessage `json:"data"`
+	}
+
+	var responses []r
+
+	if err := json.Unmarshal(data, &responses); err != nil {
+		return errors.Wrap(err, "fail to decode actions")
+	}
+
+	actions := make([]Action, 0, len(responses))
+	var action Action
+
+	for _, response := range responses {
+		switch response.ActionType {
+		case "POLICY_CHANGE":
+			action = &PolicyChangeAction{
```

Comment:
i wonder if it's not possible keep it as `[]Action` and make decisions based on `ActionType`. Lazy casting basically

Reply:
Yeah, I hard **mixed** feeling went implementing this, I had two ideas:

1. Make it generic, extract the minimal required fields for the action ( type / action ID) and puts everything else in a `json.RawMessage` field so it can be lazily extracted later one.
2. Implement a base type + concrete type on top of it.


I've gone with the second option because I wanted to keep the serialization in a single place and theses are just data object no logic should be added to them. If new behavior is required, we either:

1. Create a new type / type cast to a new type.
2. Create a function that accepts the struct.


Now, I diverge on that with the Event, as you may have noticed I've created a `SerializableEvent` interface that implements a minimum public interface and expect the user to correctly serialize the type to the JSON. I did this because, an event from Agent doesn't have at the moment a finite number of possibilities.